File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
# 입력 문자열 리스트
input_strings = []
while True:
    input_str = input()  # 입력 문자열
    if input_str == '.':  # 입력 종료 조건
        break
    input_strings.append(input_str)

for input_str in input_strings:
    check = []  # 괄호를 체크하기 위한 리스트
    answer = 'yes'  # 기본 답변은 'yes'

    for s in input_str:
        if s == '(' or s == '[':  # 열린 괄호가 들어오면 체크 리스트에 추가
            check.append(s)
        elif s == ')':  # 닫힌 소괄호 처리
            if not check or check.pop() != '(':  # 열린 괄호가 없거나 잘못된 짝인 경우
                answer = 'no'
                break
        elif s == ']':  # 닫힌 대괄호 처리
            if not check or check.pop() != '[':  # 열린 괄호가 없거나 잘못된 짝인 경우
                answer = 'no'
                break

    if len(check) > 0:  # 괄호 검사가 끝난 후에도 리스트에 괄호가 남아있는 경우
        answer = 'no'
        
    print(answer)  # 결과 출력

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        while (true) {
            String inputStr = sc.nextLine();
            
            // 종료 조건: 입력이 '.'일 경우
            if (inputStr.equals(".")) {
                break;
            }
            
            // 괄호 체크를 위한 스택
            Stack<Character> check = new Stack<>();
            String result = "yes";
            
            for (int i = 0; i < inputStr.length(); i++) {
                char c = inputStr.charAt(i);
                
                if (c == '(' || c == '[') {
                    check.push(c);  // 열린 괄호가 나오면 스택에 추가
                } else if (c == ')') {
                    if (check.isEmpty() || check.pop() != '(') {  // 열린 괄호가 없거나, 마지막에 들어간 괄호가 소괄호가 아닐 경우
                        result = "no";
                        break;
                    }
                } else if (c == ']') {
                    if (check.isEmpty() || check.pop() != '[') {  // 열린 괄호가 없거나, 마지막에 들어간 괄호가 대괄호가 아닐 경우
                        result = "no";
                        break;
                    }
                }
            }
            
            // 스택에 남아있는 괄호가 있으면 잘못된 괄호 구조
            if (!check.isEmpty()) {
                result = "no";
            }
            
            // 결과 출력
            System.out.println(result);
        }
        
        sc.close();
    }
}
